coordinator/src/notification_service.rs

```diff
@@ -130,3 +144,205 @@ async fn send_notification<'a>(
+            &[
+                position_far_from_expiring,
+                position_soon_to_expire.clone(),
+                ancient_position,
+            ],
+            &notification_sender,
+        )
+        .await;
+
+        let received_notifications = receive_all_notifications(&mut notification_receiver);
+
+        assert_eq!(received_notifications.len(), 1);
+
+        let notification = received_notifications.first().unwrap();
+        assert_eq!(
+            notification.notification_kind,
+            NotificationKind::PositionSoonToExpire
+        );
+        assert_eq!(notification.user_fcm_token, position_soon_to_expire.1);
+    }
+
+    #[tokio::test]
+    async fn send_only_recently_expired_notifications() {
+        init_tracing_for_test();
+
+        let position_far_from_expiring = far_from_expiry_position();
+        let position_just_expired = just_expired_position();
+        let ancient_position = ancient_expired_position(); // too old to send notification
+        let (notification_sender, mut notification_receiver) = mpsc::channel(100);
+
+        send_expiry_notifications_if_applicable(
+            &[
+                position_far_from_expiring,
+                position_just_expired.clone(),
+                ancient_position,
+            ],
+            &notification_sender,
+        )
+        .await;
+
+        let received_notifications = receive_all_notifications(&mut notification_receiver);
+
+        assert_eq!(received_notifications.len(), 1);
+
+        let notification = received_notifications.first().unwrap();
+        assert_eq!(
+            notification.notification_kind,
+            NotificationKind::PositionExpired
+        );
+        assert_eq!(notification.user_fcm_token, position_just_expired.1);
+    }
+}
+
+/// Load all recent positions with the DB with associated fcm tokens and send push
+/// notification about expiring/expired positions if needed.
+pub async fn query_and_send_position_notifications(
+    conn: &mut PgConnection,
+    notification_sender: &mpsc::Sender<Notification>,
+) -> Result<()> {
+    let users = user::all(conn)?;
+
+    let positions_with_fcm_tokens = positions::Position::get_all_positions_with_expiry_after(
+        conn,
+        OffsetDateTime::now_utc() - time::Duration::hours(1),
+    )?
+    .into_iter()
+    // Join positions with users to add the FCM tokens.
+    // Filter out positions that don't have a FCM token stored in the users
+    // table which is with them.
+    // This can be done at the DB level if it ever becomes a performance issue.
+    .filter_map(|p| {
+        let maybe_fcm_token = users
+            .iter()
+            .find(|u| u.pubkey == p.trader.to_string())
+            .map(|u| FcmToken(u.fcm_token.clone()));
+
+        if let Some(fcm_token) = maybe_fcm_token {
+            Some((p, fcm_token))
+        } else {
+            tracing::warn!(?p, "No FCM token for position");
+            None
+        }
```

Comment:
:wrench: This can just be a `map` (or a `find_map` altogether).

Reply:
the reason I unfolded is that I wanted to print out the warning instead of just ignoring the ones that didn't have a FCM token.

if you have a solution that does it even neater with that requirement, please share, as I always am keen on improving my iterator-fu 🙏☺️

Reply:
Fair enough. I think you could have done that within the `find_map`, but this code has diverged so I can't really give it a go myself anymore!